File: src/cartservice/src/ServiceNameInterceptor.cs
using Grpc.Core;
using Grpc.Core.Interceptors;
using OpenTelemetry.Trace;

public class ServiceNameInterceptor : Interceptor
{
    private static readonly string attributeNetPeerName = "net.peer.name";
    public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
        TRequest request,
        ServerCallContext context,
        UnaryServerMethod<TRequest, TResponse> continuation)
    {
        var clientServiceName = context.RequestHeaders?.GetValue("X-Service-Name");
        var span = Tracer.CurrentSpan;

        if (!string.IsNullOrEmpty(clientServiceName))
        {
            span.SetAttribute(attributeNetPeerName, clientServiceName);
        }

        return await continuation(request, context);
    }
}

File: src/adservice/src/main/java/oteldemo/ServiceNameInterceptor.java
package oteldemo;

import io.grpc.*;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.context.Scope;

public class ServiceNameInterceptor implements ServerInterceptor {

    private static final String ATTRIBUTE_NET_PEER_NAME = "net.peer.name";

    private static final Metadata.Key<String> X_SERVICE_NAME_KEY = Metadata.Key.of("X-Service-Name",
            Metadata.ASCII_STRING_MARSHALLER);

    @Override
    public <R, S> ServerCall.Listener<R> interceptCall(ServerCall<R, S> call,
            Metadata headers,
            ServerCallHandler<R, S> next) {
        String clientServiceName = headers.get(X_SERVICE_NAME_KEY);
        Span span = Span.current();
        if (clientServiceName != null && !clientServiceName.isEmpty()) {
            span.setAttribute(ATTRIBUTE_NET_PEER_NAME, clientServiceName);
        }

        try (Scope scope = span.makeCurrent()) {
            return next.startCall(call, headers);
        } catch (Exception e) {
            span.setStatus(StatusCode.ERROR, "Exception while processing call");
            throw e;
        } finally {
            span.end();
        }
    }
}
